File: gmb_review_link_test.py
import requests
import re

def get_review_url(gmb_share_url):
    """Convert a GMB share URL to a direct Write Review URL."""
    try:
        # Step 1: Resolve short URL to full Google Maps URL
        response = requests.head(gmb_share_url, allow_redirects=True, timeout=10)
        resolved_url = response.url
        print(f"Resolved URL: {resolved_url}\n")

        # Step 2: Extract HEX CID from !1s parameter
        match = re.search(r'!1s(0x[0-9a-fA-F]+:0x[0-9a-fA-F]+)', resolved_url)
        if not match:
            print("ERROR: HEX CID not found in URL")
            return None

        hex_cid = match.group(1)
        print(f"HEX CID: {hex_cid}\n")

        # Step 3: Build Write Review URL
        review_url = f"https://www.google.com/maps/place//data=!4m3!3m2!1s{hex_cid}!12e1"
        return review_url

    except Exception as e:
        print(f"ERROR: {e}")
        return None


# ============ TEST ============
gmb_url = "https://maps.app.goo.gl/yvBRbChDD1dDjwUe6"

print("=" * 60)
print("GMB Share URL → Write Review Link Generator")
print("=" * 60)
print(f"\nInput URL: {gmb_url}\n")

review_link = get_review_url(gmb_url)

if review_link:
    print(f"Write Review URL:\n{review_link}")
else:
    print("Failed to generate review link")

File: gmb_review_link_test.js
const https = require('https');
const http = require('http');
const url = require('url');

function resolveRedirects(inputUrl, maxRedirects = 10) {
    return new Promise((resolve, reject) => {
        let redirectCount = 0;

        function follow(currentUrl) {
            if (redirectCount >= maxRedirects) {
                return reject(new Error('Too many redirects'));
            }

            const parsed = new URL(currentUrl);
            const client = parsed.protocol === 'https:' ? https : http;

            const req = client.request(currentUrl, { method: 'HEAD' }, (res) => {
                if ([301, 302, 303, 307, 308].includes(res.statusCode) && res.headers.location) {
                    redirectCount++;
                    let nextUrl = res.headers.location;
                    // Handle relative URLs
                    if (!nextUrl.startsWith('http')) {
                        nextUrl = new URL(nextUrl, currentUrl).href;
                    }
                    follow(nextUrl);
                } else {
                    resolve(currentUrl);
                }
            });

            req.on('error', reject);
            req.end();
        }

        follow(inputUrl);
    });
}

function extractHexCID(resolvedUrl) {
    const match = resolvedUrl.match(/!1s(0x[0-9a-fA-F]+:0x[0-9a-fA-F]+)/);
    return match ? match[1] : null;
}

function buildReviewUrl(hexCID) {
    return `https://www.google.com/maps/place//data=!4m3!3m2!1s${hexCID}!12e1`;
}

// ============ TEST ============
async function main() {
    const gmbUrl = "https://maps.app.goo.gl/yvBRbChDD1dDjwUe6";

    console.log("=".repeat(60));
    console.log("GMB Share URL -> Write Review Link Generator");
    console.log("=".repeat(60));
    console.log(`\nInput URL: ${gmbUrl}\n`);

    try {
        // Step 1: Resolve
        const resolvedUrl = await resolveRedirects(gmbUrl);
        console.log(`Resolved URL:\n${resolvedUrl}\n`);

        // Step 2: Extract HEX CID
        const hexCID = extractHexCID(resolvedUrl);
        if (!hexCID) {
            console.log("ERROR: HEX CID not found in URL");
            return;
        }
        console.log(`HEX CID: ${hexCID}\n`);

        // Step 3: Build review URL
        const reviewUrl = buildReviewUrl(hexCID);
        console.log(`Write Review URL:\n${reviewUrl}`);
    } catch (err) {
        console.error(`ERROR: ${err.message}`);
    }
}

main();
